pkg/chunkenc/memchunk.go

```diff
@@ -1228,10 +1253,114 @@ func (si *bufferedIterator) moveNext() (int64, []byte, bool) {
 				continue
 			}
 			si.err = err
-			return 0, nil, false
+			return 0, nil, nil, false
 		}
 	}
-	return ts, si.buf[:lineSize], true
+
+	if si.format < chunkFormatV4 {
+		return ts, si.buf[:lineSize], nil, true
+	}
+
+	// TODO: This is pretty similar to how we read the line size, and the metadata name and value sizes
+	//       Maybe we can extract it to a separate function and reuse it?
+	var labelsWidth, nLabels int
+	for labelsWidth == 0 { // Read until we have enough bytes for the labels.
+		n, err := si.reader.Read(si.readBuf[si.readBufValid:])
+		si.readBufValid += n
+		if err != nil {
+			if err != io.EOF {
+				si.err = err
+				return 0, nil, nil, false
+			}
+			if si.readBufValid == 0 { // Got EOF and no data in the buffer.
+				return 0, nil, nil, false
+			}
+			if si.readBufValid == lastAttempt { // Got EOF and could not parse same data last time.
+				si.err = fmt.Errorf("invalid data in chunk")
+				return 0, nil, nil, false
+			}
+		}
+		var l uint64
+		l, labelsWidth = binary.Uvarint(si.readBuf[:si.readBufValid])
+		nLabels = int(l)
+		lastAttempt = si.readBufValid
+	}
+
+	// Shift down what is still left in the fixed-size read buffer, if any.
+	si.readBufValid = copy(si.readBuf[:], si.readBuf[labelsWidth:si.readBufValid])
+
+	// If not enough space for the labels, create a new buffer slice and put the old one back in the pool.
+	if nLabels*2 > cap(si.metaLabelsBuf) {
+		if si.metaLabelsBuf != nil {
+			for i := range si.metaLabelsBuf {
+				BytesBufferPool.Put(si.metaLabelsBuf[i])
+			}
+		}
+		si.metaLabelsBuf = make([][]byte, nLabels*2)
```

Comment:
Done with [0fa7568](https://github.com/grafana/loki/pull/9700/commits/0fa7568aebd4681e804d4865143b58b6a3cc5e48).

I set a maximum size of 256, i.e. max of 128 meta labels. That should be more than enough, but IIUC, we should now set a hard limit on this and refuse payloads with entries containing more than 128 meta-labels. Alternatively, instead of adding a limit and if we ever hit this maximum, we can increase the pool size. Wdyt?